File: AIReaderKindle/src/Domain/Formats/DictionaryConverter.cpp
#include "DictionaryConverter.hpp"

#include "../../Support/Files.hpp"
#include "DSLDictionaryReader.hpp"
#include "DelimitedDictionaryReader.hpp"
#include "StarDictReader.hpp"
#include "XDXFDictionaryReader.hpp"

namespace DictionaryConverter {

std::optional<DictionaryImportInfo> read(const DictionarySource& source, const DictionaryEntrySink& entry, std::string* error) {
    std::optional<DictionaryImportInfo> info;
    switch (source.format) {
    case DictionaryFormat::Native:
        if (error) *error = Files::baseName(source.main) + " is not a dictionary format this app converts.";
        return std::nullopt;
    case DictionaryFormat::Delimited: info = DelimitedDictionaryReader::read(source, entry); break;
    case DictionaryFormat::Xdxf: info = XDXFDictionaryReader::read(source, entry); break;
    case DictionaryFormat::Dsl: info = DSLDictionaryReader::read(source, entry); break;
    case DictionaryFormat::StarDict: return StarDictReader::read(source, entry, error);
    }
    if (!info && error) *error = Files::baseName(source.main) + " could not be read.";
    return info;
}

}  // namespace DictionaryConverter

File: AIReader/AIReader/Services/Dictionary/DictionaryConverter.swift
import Foundation

/// Turns a dictionary in one of the formats readers hand around into the pack
/// this app searches.
enum DictionaryConverter {
    static func convert(_ source: DictionarySource, to destination: URL) throws -> DictionaryImportInfo {
        let writer = try DictionaryPackWriter(creating: destination)
        let info = try read(source) { try writer.add($0) }
        try writer.finish(info)
        return info
    }

    private static func read(
        _ source: DictionarySource,
        entry: (DictionaryImportEntry) throws -> Void
    ) throws -> DictionaryImportInfo {
        switch source.format {
        case .native:
            throw DictionaryReadError.unrecognized(source.main.lastPathComponent)
        case .delimited:
            try DelimitedDictionaryReader.read(source, entry: entry)
        case .xdxf:
            try XDXFDictionaryReader.read(source, entry: entry)
        case .dsl:
            try DSLDictionaryReader.read(source, entry: entry)
        case .starDict:
            try StarDictReader.read(source, entry: entry)
        }
    }
}
